### PR TITLE
RSDK-2014 & RSDK-2027 & RSDK- 2025 - Implement cartographer model + integration tests & set up integration tests in CI

### PR DESCRIPTION
**Pull request by kkufieta**
Tickets:
* [RSDK-2014 - Implement cartographer model](https://viam.atlassian.net/browse/RSDK-2014)
* [RSDK-2025 - Implement cartographer integration tests](https://viam.atlassian.net/browse/RSDK-2025)
* [RSDK-2027 - Set up cartographer integration tests automation in CI](https://viam.atlassian.net/browse/RSDK-2027)

This PR is an accumulation of PRs to make the changes more digestible:
* The first step was to copy/paste the relevant RDK code (mostly builtin.go & related files)
* The following PRs are merged into this branch/PR:
    * https://github.com/viamrobotics/viam-cartographer/pull/14
    * https://github.com/viamrobotics/viam-cartographer/pull/20
    * https://github.com/viamrobotics/viam-cartographer/pull/22
    * https://github.com/viamrobotics/viam-cartographer/pull/24
    * https://github.com/viamrobotics/viam-cartographer/pull/28

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -11,7 +11,6 @@
 *.swp
 
 # Generated files and folders
-protoc-gen-grpc-cpp
```
Comment (EshaMaharishi) [inline .gitignore:14]: Was this auto-generated or any context on why it's removed?

File: integration_test.go [hunk 2]
```diff
@@ -0,0 +1,389 @@
+package viamcartographer_test
```
Comment (EshaMaharishi) [inline integration_test.go:1]: For this and the files below, would you mind adding a comment that links to which file(s) in rdk they're largely copied from?

**Review by tessavitabile — APPROVED** [review]
LGTM mod comment

(Though my understanding is this will not be merged until subsequent work on this branch)

File: internal/grpchelper/grpchelper.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
+// Package grpchelper implements helper functions to be used with slam service grpc clients
```
[the remark this answers is not in the document]
Comment (kkufieta) [reply internal/grpchelper/grpchelper.go:1]: Yes, we made a copy of it. We could extract it into the slam repo & export it - my preferred option - or leave a copy here. Please let me know what your thoughts are.

File: .gitignore (hunk 1 above)
Comment (kkufieta) [reply .gitignore:14]: Sorry - it's a remnant (dead code).

File: integration_test.go (hunk 2 above)
Comment (kkufieta) [reply integration_test.go:1]: Good idea! 

integration_test.go was formerly [cartographer_int_test.go](https://github.com/viamrobotics/rdk/blob/main/services/slam/builtin/cartographer_int_test.go)

File: internal/grpchelper/grpchelper.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (kkufieta) [reply internal/grpchelper/grpchelper.go:1]: For Esha: This file is a 1:1 copy from [slam/internal/grpchelper/grpchelper.go](https://github.com/viamrobotics/rdk/blob/main/services/slam/internal/grpchelper/grpchelper.go). We decided to do it this way due to our discussion yesterday where it seemed like we shouldn't keep extracting more code into the slam repo. My preferred choice is to extract this into the slam repo.

File: internal/testhelper/testhelper.go
Comment (kkufieta) [inline internal/testhelper/testhelper.go:1]: This is only a part of [slam/internal/testhelper/testhelper.go](https://github.com/viamrobotics/rdk/blob/main/services/slam/internal/testhelper/testhelper.go)

File: viam-cartographer.go
Comment (kkufieta) [inline viam-cartographer.go:1]: viam-cartographer.go --> formerly [builtin.go](https://github.com/viamrobotics/rdk/blob/main/services/slam/builtin/builtin.go)

File: viam-cartographer_test.go
Comment (kkufieta) [inline viam-cartographer_test.go:1]: viam-cartographer_test.go --> formerly [builtin_test.go](https://github.com/viamrobotics/rdk/blob/main/services/slam/builtin/builtin_test.go)

File: internal/testhelper/testhelper.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
+// Package testhelper implements a slam service definition with additional exported functions for
```
Comment (EshaMaharishi) [inline internal/testhelper/testhelper.go:1]: [optional] Should we remove the "with additional exported functions for the purpose of testing"? (Or maybe we will add more functions later?)

File: internal/grpchelper/grpchelper.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (EshaMaharishi) [reply internal/grpchelper/grpchelper.go:1]: I don't have an opinion yet, but agree it doesn't seem great to have 3 copies of these functions.

**Review by EshaMaharishi — APPROVED** [review]
LGTM with optional comment

File: internal/testhelper/testhelper.go (hunk 4 above)
Comment (kkufieta) [reply internal/testhelper/testhelper.go:1]: Yes, I agree. I will do that in subsequent PRs!

Comment (EshaMaharishi) [reply internal/testhelper/testhelper.go:1]: Sounds good, ty!

File: viam-cartographer.go [hunk 5]
```diff
@@ -0,0 +1,988 @@
...
+		}
+
+		pInFrame = referenceframe.ProtobufToPoseInFrame(resp.Pose)
+		returnedExt = resp.Extra.AsMap()
+	}
+
+	// TODO DATA-531: https://viam.atlassian.net/jira/software/c/projects/DATA/boards/30?modal=detail&selectedIssue=DATA-531
+	// Remove extraction and conversion of quaternion from the extra field in the response once the Rust
+	// spatial math library is available and the desired math can be implemented on the orbSLAM side
+	if val, ok := returnedExt["quat"]; ok {
```
Comment (zaporter-work) [inline viam-cartographer.go:299]: [nit] should this have been refactored because we extracted the `CheckQuaternionFromClientAlgo()`?

**Review by zaporter-work — APPROVED** [review]
LGTM mod quat refactor

File: internal/grpchelper/grpchelper.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (kkufieta) [reply internal/grpchelper/grpchelper.go:1]: We decided to make grpchelper.go public by moving it out of the internal folder, and import it into the orbslam & cartographer repos: https://viam.atlassian.net/browse/RSDK-2354

**Comment by kkufieta** [comment]
@zaporter-work @tessavitabile @EshaMaharishi this PR should be ready to be merged. The only thing I have changed outside of the merged PRs is adding the `-race` flag to the integration tests:

* https://github.com/viamrobotics/viam-cartographer/pull/15/commits/9b432fa34ce04aa1cc9646fd6ea1efccd37bc7bd

If that looks good to you, I'll go ahead and merge.

**Review by EshaMaharishi — APPROVED** [review]

**Pull request merged**